### PR TITLE
Add options to change braille cursor visibility and shape

### PR DESCRIPTION
**Pull request by dkager**
Related issue: #5198

The option to toggle the cursor on/off is something I'd really like to have. The option to change the cursor shape was suggested by someone else but seems useful.

Also contains a minor refactor: define a constant `SELECTION_SHAPE` instead of computing it over and over again when marking the selection.

### CODE REVIEW

**Comment by jcsteh** [comment]
Do you still have the t5410_t5198 branch you mentioned [here](https://github.com/nvaccess/nvda/issues/5410#issuecomment-155343904) somewhere? I decided to do #5410 first, as it makes this change a little cleaner, but that unfortunately does break this pull request.

**Comment by dkager** [comment]
I don’t have it up anywhere but I can make it come back or do the unspeakable and rebase on the commit where #5410 was incubated.

**Comment by dkager** [comment]
I put the branch back up [here](https://github.com/dkager/nvda/tree/t5410_t5198). Also includes another commit to fix what I think was an oversight on my end.

**Comment by jcsteh** [comment]
Heh, yeah, I seem to have overlooked that error too when reviewing.

**Comment by jcsteh** [comment]
I'm providing some code review here for reference, but I'll apply it myself, since I'm going to have to use a different branch and merging is going to get messy.

File: source/braille.py [hunk 1]
```diff
@@ -352,8 +352,16 @@
 	controlTypes.STATE_CHECKED: _("( )"),
 }
 
-DOT7 = 64
-DOT8 = 128
+#: Cursor shapes
+CURSOR_SHAPES = (
+	# Translators: The description of a cursor shape.
```
Comment (jcsteh) [inline source/braille.py]: nit: The translators comment should be specific and say "braille cursor shape", as the translator might not know this is related to braille (they may not see/shouldn't have to figure this out from the source file name).

File: user_docs/en/userGuide.t2t [hunk 2]
```diff
@@ -888,9 +888,15 @@ If input is not supported on a display which does have a braille keyboard, this
 ==== Expand to computer braille for the word at the cursor ====
 This option allows the word that is under the cursor to be displayed in non-contracted computer braille.
 
-==== Cursor Blink Rate ====
+==== Show Cursor ====
+This option allows the braille cursor to be turned on and off. It applies to all cursors including the selection indicator.
```
Comment (jcsteh) [inline user_docs/en/userGuide.t2t]: nit: In the User Guide, each sentence should generally be on a separate line to make life easier for translators. Same applies to a later paragraph.

File: user_docs/en/userGuide.t2t [hunk 3]
```diff
@@ -888,9 +888,15 @@ If input is not supported on a display which does have a braille keyboard, this
...
 
-==== Cursor Blink Rate ====
+==== Show Cursor ====
+This option allows the braille cursor to be turned on and off. It applies to all cursors including the selection indicator.
+
+==== Cursor Blink Rate (ms) ====
 This option is a numerical field that allows you to change the blink rate of the cursor in milliseconds.
 
+==== Cursor Shape ====
+This option allows you to choose the shape of the braille cursor. Shapes are defined as dot patterns. The selection indicator is not affected by this option.
```
Comment (jcsteh) [inline user_docs/en/userGuide.t2t]: nit: I think "Shapes are defined as dot patterns" is a bit confusing for some, even though I know exactly what you mean. I've just put dot pattern in parentheses like so:

> This option allows you to choose the shape (dot pattern) of the braille cursor.

File: source/braille.py [hunk 4]
```diff
@@ -920,13 +928,13 @@ def update(self):
...
 		if self._selectionStart is not None:
-			# Mark the selection with dots 7 and 8.
+			# Mark the selection.
 			if self._selectionEnd >= len(self.rawText):
 				brailleSelEnd = len(self.brailleCells)
 			else:
 				brailleSelEnd = self.rawToBraillePos[self._selectionEnd]
 			for pos in xrange(self.rawToBraillePos[self._selectionStart], brailleSelEnd):
-				self.brailleCells[pos] |= DOT7 | DOT8
+				self.brailleCells[pos] |= SELECTION_SHAPE
```
Comment (jcsteh) [inline source/braille.py]: The documentation for Show cursor says it affects the selection indicator, but it doesn't. I assume the documentation reflects your intent and the code does not? :)

**Comment by jcsteh** [comment]
I squashed, applied code review and pushed to [the nvaccess i5198 branch](/nvaccess/nvda/tree/i5198). For now, the selection is disabled with the cursor, but this might change depending on the discussion in #5198. If it doesn't, a fix will be needed both here and in #5410 to make it work when toggled via the script.

**Comment by dkager** [comment]
OK so i5198 is the most up-to-date? Can the t5410_t5198 merge branch be ignored now? Just trying to figure out where further work and testing should happen.

**Comment by jcsteh** [comment]
i5198 is the most up to date, yes. It is based on our i5410 branch. 
However, I'd still prefer we kept changes specific to #5410 in the i5410 
branch. I realise there is interdependence, but we want separability if 
we choose to take #5410 but defer #5198 or similar.

**Comment by dkager** [comment]
Updated my t5198 branch for consistency.

**Comment by jcsteh** [comment]
Weird. The pull request doesn't reflect your latest commit, but my understanding of pull requests is that they should update with any further commits to the branch.

**Comment by dkager** [comment]
It does show here, in 5a32192, which is just a copy of i5198 minus the unrelated stuff from t5410. A bit redundant info now but there you go.

**Comment by jcsteh** [comment]
Bah. Just me being stupid; I forgot GitHub splits commits into multiple tables for different dates. Sorry.

**Pull request closed without merging**